query.go

```diff
@@ -251,32 +262,37 @@ func (dht *IpfsDHT) constructLookupResult(queries []*query, target kb.ID) *looku
 }
 
 type queryUpdate struct {
-	seen        []peer.ID
+	cause       peer.ID
+	heard       []peer.ID
 	queried     []peer.ID
 	unreachable []peer.ID
 }
 
-func (q *query) runWithGreedyParallelism() {
+func (q *query) runAsync() {
 	pathCtx, cancelPath := context.WithCancel(q.ctx)
 	defer cancelPath()
 
 	alpha := q.dht.alpha
 
 	ch := make(chan *queryUpdate, alpha)
-	ch <- &queryUpdate{seen: q.seedPeers}
+	ch <- &queryUpdate{heard: q.seedPeers}
 
 	for {
+		var cause peer.ID
 		select {
 		case update := <-ch:
-			q.updateState(update)
+			q.updateState(pathCtx, update)
+			cause = update.cause
```

Comment:
This is ok, but it means that if querying A gives us A1, A2, A3 and then we receive B1 and B2 from B we change the cause to B. In the future when we query A1 the cause will be B instead of A.

We can probably back calculate that A1 came from A, but we may want to include that information here as well.

Note: In the future we may allow transitioning state from Unreachable to Heard based on getting a new multiaddress via signed peer records. If so then if we're back calculating we may need to take into account that A gave us A1 with a bad address, but that F gave us A1 with the address we actually used.

Reply:
Yes, as we spoke: In the future, we can attach the "source" of a peer (where we heard about it) to the state.
This is a different type of information than the cause. The cause tells is which peer's response triggered our next action.

Reply:
we now have a source